File: py/141.py
# Author: Andrew Hamara

# Solution to leetcode problem 141. Linked List Cycle

# Use hare/tortoise implementation, the hare will always catch the tortoise!

class Solution:
    def hasCycle(self, head:Optional[ListNode]) -> bool:
        hare = head
        tortoise = head

        while hare and hare.next:
            hare = hare.next.next
            tortoise = tortoise.next

            if hare == tortoise:
                return True
        return False

File: go/141.go
// Author: Andrew Hamara

// Solution to leetcode problem 141. Linked List Cycle

func hasCycle(head *ListNode) bool {

    hare := head
    tortoise := head

    for hare != nil && tortoise != nil {
        hare = hare.next.next
        tortoise = tortoise.next

        if hare == tortoise {
            return true
        }
    }
    return false
}
